### PR TITLE
Merge v1.16.1 into master

### PR DESCRIPTION
**Pull request by bnoordhuis**
Last update was in July so it was about time again.

CI: https://ci.nodejs.org/job/libuv-test-commit/572/

### CODE REVIEW

**Comment by santigimeno** [comment]
There seem to be compilation issues on Windows

**Comment by santigimeno** [comment]
From https://ci.nodejs.org/job/libuv-test-commit-windows/663/nodes=win2016-vs2015/console
```
test\test-fs-copyfile.c(71): error C2065: 'O_WRONLY': undeclared identifier [c:\workspace\libuv-test-commit-windows\nodes\win2016-vs2015\run-tests.vcxproj]
test\test-fs-copyfile.c(71): error C2065: 'O_CREAT': undeclared identifier [c:\workspace\libuv-test-commit-windows\nodes\win2016-vs2015\run-tests.vcxproj]
test\test-fs-copyfile.c(71): error C2065: 'O_TRUNC': undeclared identifier [c:\workspace\libuv-test-commit-windows\nodes\win2016-vs2015\run-tests.vcxproj]
```

**Comment by bnoordhuis** [comment]
Missing an `#include <fcntl.h>`.  Updated, new CI: https://ci.nodejs.org/job/libuv-test-commit/576/

**Comment by bnoordhuis** [comment]
Windows test failures in test-fs.c and test-fs-copyfile.c.  CI of master to compare against: https://ci.nodejs.org/job/libuv-test-commit/577/

**Comment by bnoordhuis** [comment]
Looks like the Windows failure is the same issue as the one described in #1637.
 
<sup>Note to self: failing perma-build for posterity: https://ci.nodejs.org/job/libuv-test-commit-windows/667/nodes=win10-vs2015/</sup>

**Comment by bnoordhuis** [comment]
Synced with v1.x, includes changes from #1639: https://ci.nodejs.org/job/libuv-test-commit/588/

**Comment by santigimeno** [comment]
@bnoordhuis this [commit](https://github.com/santigimeno/libuv/commit/07187960221ee717509fb498998857cee53add05) seems to fix the issue though I agree with your comment [here](https://github.com/libuv/libuv/pull/1166#discussion_r151144762) about the changing of semantics of `uv_fs_open()` return value. I think, at least, it should have been documented.
Also I have added a [commit](https://github.com/santigimeno/libuv/commit/97403d1430a5b6cd3bc44b353d8937ef8146f964) to fix a warning on windows in the same branch: https://github.com/santigimeno/libuv/commits/santi_merge_master

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/642/

**Comment by saghul** [comment]
Thanks for doing this Ben! I dropped the ball on this one :-(

File: include/uv/pthread-barrier.h [hunk 1]
```diff
@@ -18,15 +18,45 @@ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
...
 #include <pthread.h>
+#if !defined(__MVS__)
+#include <semaphore.h> /* sem_t */
+#endif
 
 #define PTHREAD_BARRIER_SERIAL_THREAD  0x12345
+#define UV__PTHREAD_BARRIER_FALLBACK   1
+
+/*
+ * To maintain ABI compatibility with
```
Comment (saghul) [inline include/uv/pthread-barrier.h]: I believe this needs to go on master, right?

**Comment by cjihrig** [comment]
With @santigimeno's changes, would we be able to get this landed soon? I'd like to start working on a v2 release within the next month if possible.

File: include/uv/pthread-barrier.h (hunk 1 above)
Comment (saghul) [reply include/uv/pthread-barrier.h]: By go I mean get removed.

**Comment by cjihrig** [comment]
Is this stuck?

If it is, would it make sense to cut a new v2 branch and cherry-pick over the commits that have landed on master (branch-diff has that at about 50 commits, excluding merge commits, but I assume at least some of them are rather involved)?

**Comment by bnoordhuis** [comment]
Stuck on making it build on Windows.  Some of the changes in master are pretty involved, yes.

**Comment by santigimeno** [comment]
@bnoordhuis there were a [couple of commits](https://github.com/libuv/libuv/pull/1630#issuecomment-354427914) that I think fixed the Windows issues. Did you see them?

**Comment by santigimeno** [comment]
This can be closed as https://github.com/libuv/libuv/pull/1753 has already landed.

**Pull request closed without merging**